### DISCUSSION TITLE
Integration with Hakyll

### DISCUSSION THREAD

**Discussion by livmackintosh** [srid/emanote#195] [Q&A]
I have an existing Hakyll site and a neuron project. I want to upgrade neuron to emanote and then integrate the compilation with my Hakyll rules. Is there any way to do this?

I don't want to use Nix or Docker because I'm using stack and cabal and want to keep it that way.

Olivia

**Comment by srid** [comment]
You should be able to use Emanote as a Haskell library, but it requires a few things:

1. Add a `library` stanza to the cabal file (but without breaking Nix & bin/run ghcid workflow)
2. Add cabal.project or stack.yaml containing the source overrides corresponding to the Haskell packages in flake.nix (and flake.lock) https://github.com/srid/emanote/blob/fd2d1cf401d04ac51adb4fbdb6248c89057fc133/flake.nix#L55-L59
3. Expose any endpoints for one-off building of files (that you can hook in your Hakyll code).

(1) and (3) should be part of https://github.com/srid/emanote/issues/188 - I don't plan on doing this any time soon, but if you are interested in contributing PRs I'd be happy to review.

For (3), you could call this code: https://github.com/srid/emanote/blob/fd2d1cf401d04ac51adb4fbdb6248c89057fc133/src/Main.hs#L17-L18 

But with `cli` set to `Generate <path>`, https://github.com/srid/ema/blob/8d4ba79adf0dc5f42c946b5b511575e2da6014c6/src/Ema/CLI.hs#L15

---

By the way, Emanote uses https://ema.srid.ca/ under the hood, which like Hakyll is also for static site generation.

**Reply by livmackintosh** [reply]
Hmm. I'm starting to feel that this project is doing way too much and I'd be better off ditching it and writing some rules to generate the backlinks. Then Pandoc/Hakyll can handle the rest. But I'll bare this in mind. Seems good if you want to "adopt the whole ecosystem" but it feels like a huge monolith. I don't want to use Nix or Docker and I don't want or need to use VSCode integrations. I just want to be able to pass my zettelkasten into a function and get a compiled site.

Olivia

**Reply by srid** [reply]
A general goal of mine is to release the individual features to Hackage, separate from Emanote.  For example, https://hackage.haskell.org/package/pandoc-link-context (which Emanote uses) is already isolated - and I imagine you can use it in your writing-your-own approach. If you have any other suggestions for decomposing (wikilink parsing and graph building comes to mind), let me know, I'll add it to the list for 1.0.

**Comment by srid** [comment]
Just wanted to note that https://hackage.haskell.org/package/heist-extra has been extracted from Emanote. It handles the HTML templating part[^1].

Also, Emanote is [close to being released](https://github.com/EmaApps/emanote/issues/153). If there is anything else that should be extracted from Emanote and made its own library, this is the appropriate time to request! 

[^1]: See example non-emanote usage here: https://github.com/EmaApps/timedot-invoice

**Reply by srid** [reply]
Wikilink parser is now a separate library: https://hackage.haskell.org/package/commonmark-wikilink

cf. https://github.com/EmaApps/emanote/pull/367
